src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsManagedIdentitySource.cs

```diff
@@ -211,5 +220,103 @@ public static Uri GetValidatedEndpoint(
 
             return builder.Uri;
         }
+
+        public static string ImdsQueryParamsHelper(
+            RequestContext requestContext,
+            string apiVersionQueryParam,
+            string imdsApiVersion)
+        {
+            var queryParams = $"{apiVersionQueryParam}={imdsApiVersion}";
+
+            var userAssignedIdQueryParam = GetUserAssignedIdQueryParam(
+                requestContext.ServiceBundle.Config.ManagedIdentityId.IdType,
+                requestContext.ServiceBundle.Config.ManagedIdentityId.UserAssignedId,
+                requestContext.Logger);
+
+            if (userAssignedIdQueryParam != null)
+            {
+                queryParams += $"&{userAssignedIdQueryParam.Value.Key}={userAssignedIdQueryParam.Value.Value}";
+            }
+
+            return queryParams;
+        }
+
+        public static async Task<bool> ProbeImdsEndpointAsync(
+            RequestContext requestContext,
+            ImdsVersion imdsVersion)
+        {
+            string apiVersionQueryParam;
+            string imdsApiVersion;
+            string imdsEndpoint;
+            string imdsStringHelper;
+
+            switch (imdsVersion)
+            {
+                case ImdsVersion.V2:
+#if NET462
+                requestContext.Logger.Info("[Managed Identity] IMDSv2 flow is not supported on .NET Framework 4.6.2. Cryptographic operations required for managed identity authentication are unavailable on this platform. Skipping IMDSv2 probe.");
+                return false;
+#else
+                    apiVersionQueryParam = ImdsV2ManagedIdentitySource.ApiVersionQueryParam;
+                    imdsApiVersion = ImdsV2ManagedIdentitySource.ImdsV2ApiVersion;
+                    imdsEndpoint = ImdsV2ManagedIdentitySource.CsrMetadataPath;
+                    imdsStringHelper = "IMDSv2";
+                    break;
+#endif
+                case ImdsVersion.V1:
+                    apiVersionQueryParam = ApiVersionQueryParam;
+                    imdsApiVersion = ImdsApiVersion;
+                    imdsEndpoint = ImdsTokenPath;
+                    imdsStringHelper = "IMDSv1";
+                    break;
+
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(imdsVersion), imdsVersion, null);
+            }
+
+            var queryParams = ImdsQueryParamsHelper(requestContext, apiVersionQueryParam, imdsApiVersion);
+
+            var headers = new Dictionary<string, string>
```

Comment:
would be good to have some comments here like, 

`probe deliberately omits the Metadata: true header and then treats 400 as IMDS present`